### PR TITLE
Support Unix sockets for upstream

### PR DESCRIPTION
**Pull request by pwmarcz**

### CODE REVIEW

**Comment by marcinkaszynski** [comment]
Looks good enough as implementation, thanks!

It needs a test though: see `rproxy_test.go` for examples, in particular `TestProxyUplinkTLS`.  

You'll have to expose `fakeredis` via unix socket, but that shouldn't be very time consuming.  Alternatively you might put some more time to handle `unix: true` in `config.go:AddrSpec.Listen`; this won't change `fakeredis`, but will let `redis-proxy` use unix sockets for `listen` and `admin` sockets, so the test there might look just like `TestProxyUplinkTLS`, but the first proxy would publish unix socket instead of TCP with TLS.

**Comment by marcinkaszynski** [comment]
Minor: what if we named the option `network` for consistency with `net.Dial`?
`network=tcp|unix`, defaults to `tcp`.

**Comment by pwmarcz** [comment]
Back for review.

File: fakeredis/fakeredis.go [hunk 1]
```diff
@@ -10,34 +10,37 @@ package fakeredis
...
+	listener net.Listener
 
 	mu       sync.Mutex
 	shutdown bool
 	requests []*resp.Msg
 }
 
-func New(name string) *FakeRedisServer {
-	return &FakeRedisServer{name: name}
+func New(name string, network string) *FakeRedisServer {
```
Comment (marcinkaszynski) [inline fakeredis/fakeredis.go]: Minor: you can shorten that to `New(name, network string)`

File: fakeredis/fakeredis.go [hunk 2]
```diff
@@ -10,34 +10,37 @@ package fakeredis
...
+	return &FakeRedisServer{name: name, network: network}
 }
 
-func Start(name string) *FakeRedisServer {
+func Start(name string, network string) *FakeRedisServer {
```
Comment (marcinkaszynski) [inline fakeredis/fakeredis.go]: Same here, omit repeated types.

File: fakeredis/fakeredis.go [hunk 3]
```diff
@@ -59,10 +62,33 @@ func (s *FakeRedisServer) Requests() []*resp.Msg {
 
 func (s *FakeRedisServer) Run(startedChan chan struct{}) {
 	var err error
+	var tcpListener *net.TCPListener
+	var unixListener *net.UnixListener
```
Comment (marcinkaszynski) [inline fakeredis/fakeredis.go]: Why use separate variables?  It looks like if you assign directly `s.listener = net.Listen*(...)` regardless of the actual `Listen*` function used, you could skip these, and also get rid of the switch in lines 100-107.

File: rproxy/config.go [hunk 4]
```diff
@@ -67,8 +68,17 @@ func (as *AddrSpec) AsJSON() string {
 }
 
 func (as *AddrSpec) Dial() (net.Conn, error) {
+	network := "tcp"
+	if as.Network != "" {
+		network = as.Network
+	}
+	if !(network == "tcp" || network == "unix") {
+		err := errors.New("Unsupported network for dialing: " + network)
+		return nil, err
```
Comment (marcinkaszynski) [inline rproxy/config.go]: One-line it?  `return nil, errors.New(...)`

**Comment by pwmarcz** [comment]
Back for review.

**Pull request merged**